# House sales prices in King County

A project on exploratory data analysis.

Sebastian Thomas @ neue fische Bootcamp Data Science<br />
(datascience at sebastianthomas dot de)

![house][photo]

[photo]: house.jpg "House"
<div style="text-align: right">(Source: Image of <a href="https://pixabay.com/de/users/Pexels-2286921/">Pexels</a> at <a href="https://pixabay.com/">Pixabay</a>)</div>

# Part 0: Business understanding

## Data origin
The data was given to me as a project on exploratory data analysis at the neue fische Bootcamp Data Science. It is a data set that is often used in teaching and can also be found on [kaggle](www.kaggle.com). The origin of the data set is probably https://blue.kingcounty.com/Assessor/eRealProperty/default.aspx, where additional data may be found.

## Original features

The instances represent houses in King County in the State of Washington, in the Pacific Northwest region of the United States of America. The given features are as follows:

 feature           | description                                                   | type
:------------------|:--------------------------------------------------------------|:----------------------
 `'id'`            | unique identifier, equals to parcel number, **used as index** | discrete (numeric)
 `'date'`          | date when object was sold                                     | date time
 `'price'`         | sale price, **used as prediction target**                     | continuous (numeric)
 `'bedrooms'`      | number of bedrooms in house                                   | discrete (numeric)
 `'bathrooms'`     | number of bathrooms in house                                  | discrete (numeric)
 `'sqft_living'`   | square footage of interior housing living space               | continuous (numeric)
 `'sqft_lot'`      | square footage of land lot                                    | continuous (numeric)
 `'floors'`        | total floors (levels) in house                                | discrete (numeric)
 `'waterfront'`    | indicator whether house lies at a waterfront                  | boolean
 `'view'`          | indicator for views of object                                 | ordered (categorical)
 `'condition'`     | overall condition of object                                   | ordered (categorical)
 `'grade'` | overall grade given to the housing unit, based on King County grading system | ordered (categorical)
 `'sqft_above'`    | square footage of house apart from basement                   | continuous (numeric)
 `'sqft_basement'` | square footage of basement                                    | continuous (numeric)
 `'yr_built'`      | year when house was built                                     | discrete (numeric)
 `'yr_renovated'`  | year when house was renovated                                 | discrete (numeric)
 `'zipcode'`       | zipcode of area where house is positioned                     | nominal (categorical)
 `'lat'`           | latitude coordinate of object                                 | continuous (numeric)
 `'long'`          | longitude coordinate of object                                | continuous (numeric)
 `'sqft_living15'` | square footage of interior housing living space for the nearest 15 neighbors | continuous (numeric)
 `'sqft_lot15'`    | square footage of land lots of the nearest 15 neighbors       | continuous (numeric)

### Feature `'bathrooms'`

The feature `'bathrooms'` has float values. The meaning is as follows:

 value  |    description     | meaning
:------:|:------------------:|:----------------------------------------------------
 `1.00` |     full bath      | having a toilet, a sink, a shower and a tub
 `0.75` | three-quarter-bath | having only a toilet, a sink and (usually) a shower
 `0.50` |     half-bath      | having only a toilet and a sink
 `0.25` |    powder room     | having only a sink (very rare)

<div style="text-align: right">(Source: e.g. <a href="https://en.wikipedia.org/wiki/Bathroom#Variations_and_terminology">Wikipedia</a>)</div>

### Feature `'condition'`

The meaning of the feature `'condition'` is as follows:

 value | description | meaning
:-----:|:-----------:|:--------
  `1`  |    Poor     | Worn out. Repair and overhaul needed on painted surfaces, roofing, plumbing, heating and numerous functional inadequacies. Excessive deferred maintenance and abuse, limited value-in-use, approaching abandonment or major reconstruction; reuse or change in occupancy is imminent. Effective age is near the end of the scale regardless of the actual chronological age.
  `2`  |    Fair     | Badly worn. Much repair needed. Many items need refinishing or overhauling, deferred maintenance obvious, inadequate building utility and systems all shortening the life expectancy and increasing the effective age.
  `3`  |   Average   | Some evidence of deferred maintenance and normal obsolescence with age in that a few minor repairs are needed, along with some refinishing. All major components still functional and contributing toward an extended life expectancy. Effective age and utility is standard for like properties of its class and usage.
  `4`  |    Good     | No obvious maintenance required but neither is everything new. Appearance and utility are above the standard and the overall effective age will be lower than the typical property.
  `5`  |  Very Good  | All items well maintained, many having been overhauled and repaired as they have shown signs of wear, increasing the life expectancy and lowering the effective age with little deterioration or obsolescence evident with a high degree of utility.

<div style="text-align: right">(Source: <a href="https://info.kingcounty.gov/assessor/esales/Glossary.aspx?type=r">King County Government</a>)</div>

### Feature `'grade'`

The meaning of the feature `'grade'` is as follows:

  value  | meaning
:-------:|:--------
 `1`-`3` | Falls short of minimum building standards. Normally cabin or inferior structure.
   `4`   | Generally older, low quality construction. Does not meet code.
   `5`   | Low construction costs and workmanship. Small, simple design.
   `6`   | Lowest grade currently meeting building code. Low quality materials and simple designs.
   `7`   | Average grade of construction and design. Commonly seen in plats and older sub-divisions.
   `8`   | Just above average in construction and design. Usually better materials in both the exterior and interior finish work.
   `9`   | Better architectural design with extra interior and exterior design and quality.
  `10`   | Homes of this quality generally have high quality features. Finish work is better and more design quality is seen in the floor plans. Generally have a larger square footage.
  `11`   | Custom design and higher quality finish work with added amenities of solid woods, bathroom fixtures and more luxurious options.
  `12`   | Custom design and excellent builders. All materials are of the highest quality and all conveniences are present.
  `13`   | Generally custom designed and built. Mansion level. Large amount of highest quality cabinet work, wood trim, marble, entry ways etc.
  
<div style="text-align: right">(Source: <a href="https://info.kingcounty.gov/assessor/esales/Glossary.aspx?type=r">King County Government</a>)</div>

# Part 1: Data mining

We import the data set, explore it briefly and drop duplicates.

## Imports

### Modules, classes and functions

In [1]:
# data
import pandas as pd

# custom modules
from modules.ds import data_type_info

We set the options to print all columns and to print float columns with two decimals.

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:,.2f}'.format)

### Data

We import our data. The value `''`, which occurs for the features `'waterfront'` and `'yr_renovated'`, and the value `'?'`, which occurs for the feature `'sqft_basement'`, will be imported as `NaN`.

In [3]:
houses = pd.read_csv('data/King_County_House_prices_dataset.csv', index_col='id', na_values=['', '?'])

## First exploration

We check whether the import worked as expected.

In [4]:
houses

date      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id                                                                              
7129300520  10/13/2014 221,900.00         3       1.00         1180      5650   
6414100192   12/9/2014 538,000.00         3       2.25         2570      7242   
5631500400   2/25/2015 180,000.00         2       1.00          770     10000   
2487200875   12/9/2014 604,000.00         4       3.00         1960      5000   
1954400510   2/18/2015 510,000.00         3       2.00         1680      8080   
...                ...        ...       ...        ...          ...       ...   
263000018    5/21/2014 360,000.00         3       2.50         1530      1131   
6600060120   2/23/2015 400,000.00         4       2.50         2310      5813   
1523300141   6/23/2014 402,101.00         2       0.75         1020      1350   
291310100    1/16/2015 400,000.00         3       2.50         1600      2388   
1523300157  10/15/2014 325,000.00         2       0.75         1020      1076   

            floors  waterfront  view  condition  grade  sqft_above  \
id                                                                   
7129300520    1.00         nan  0.00          3      7        1180   
6414100192    2.00        0.00  0.00          3      7        2170   
5631500400    1.00        0.00  0.00          3      6         770   
2487200875    1.00        0.00  0.00          5      7        1050   
1954400510    1.00        0.00  0.00          3      8        1680   
...            ...         ...   ...        ...    ...         ...   
263000018     3.00        0.00  0.00          3      8        1530   
6600060120    2.00        0.00  0.00          3      8        2310   
1523300141    2.00        0.00  0.00          3      7        1020   
291310100     2.00         nan  0.00          3      8        1600   
1523300157    2.00        0.00  0.00          3      7        1020   

            sqft_basement  yr_built  yr_renovated  zipcode   lat    long  \
id                                                                         
7129300520           0.00      1955          0.00    98178 47.51 -122.26   
6414100192         400.00      1951      1,991.00    98125 47.72 -122.32   
5631500400           0.00      1933           nan    98028 47.74 -122.23   
2487200875         910.00      1965          0.00    98136 47.52 -122.39   
1954400510           0.00      1987          0.00    98074 47.62 -122.05   
...                   ...       ...           ...      ...   ...     ...   
263000018            0.00      2009          0.00    98103 47.70 -122.35   
6600060120           0.00      2014          0.00    98146 47.51 -122.36   
1523300141           0.00      2009          0.00    98144 47.59 -122.30   
291310100            0.00      2004          0.00    98027 47.53 -122.07   
1523300157           0.00      2008          0.00    98144 47.59 -122.30   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503  
...                   ...         ...  
263000018            1530        1509  
6600060120           1830        7200  
1523300141           1020        2007  
291310100            1410        1287  
1523300157           1020        1357  

[21597 rows x 20 columns]

The dataframe `houses` has `21597` rows and `20` columns, i.e. we have `21597` instances, one target (`'price'`) and `19` features.

We explore the current data types, the number of unique values and the number of NA values of all features.

In [5]:
data_type_info(houses)

dtype  n_unique  p_unique  n_na  p_na
date            object       372      0.02     0  0.00
price          float64      3622      0.17     0  0.00
bedrooms         int64        12      0.00     0  0.00
bathrooms      float64        29      0.00     0  0.00
sqft_living      int64      1034      0.05     0  0.00
sqft_lot         int64      9776      0.45     0  0.00
floors         float64         6      0.00     0  0.00
waterfront     float64         2      0.00  2376  0.11
view           float64         5      0.00    63  0.00
condition        int64         5      0.00     0  0.00
grade            int64        11      0.00     0  0.00
sqft_above       int64       942      0.04     0  0.00
sqft_basement  float64       303      0.01   454  0.02
yr_built         int64       116      0.01     0  0.00
yr_renovated   float64        70      0.00  3842  0.18
zipcode          int64        70      0.00     0  0.00
lat            float64      5033      0.23     0  0.00
long           float64       751      0.03     0  0.00
sqft_living15    int64       777      0.04     0  0.00
sqft_lot15       int64      8682      0.40     0  0.00

The features `'waterfront'`, `'view'`, `'sqft_basement'` and `'yr_renovated'` have NA values.

## Dropping duplicates

We investigate those instances, whose index is equal.

In [6]:
houses[houses.index.duplicated(keep=False)][:30]

date        price  bedrooms  bathrooms  sqft_living  \
id                                                                      
6021501535   7/25/2014   430,000.00         3       1.50         1580   
6021501535  12/23/2014   700,000.00         3       1.50         1580   
4139480200   6/18/2014 1,380,000.00         4       3.25         4290   
4139480200   12/9/2014 1,400,000.00         4       3.25         4290   
7520000520    9/5/2014   232,000.00         2       1.00         1240   
7520000520   3/11/2015   240,500.00         2       1.00         1240   
3969300030   7/23/2014   165,000.00         4       1.00         1000   
3969300030  12/29/2014   239,900.00         4       1.00         1000   
2231500030   10/1/2014   315,000.00         4       2.25         2180   
2231500030   3/24/2015   530,000.00         4       2.25         2180   
8820903380   7/28/2014   452,000.00         6       2.25         2660   
8820903380    1/2/2015   730,000.00         6       2.25         2660   
726049190    10/2/2014   287,500.00         3       1.00         1810   
726049190    2/18/2015   431,000.00         3       1.00         1810   
8682262400   7/18/2014   430,000.00         2       1.75         1350   
8682262400   5/13/2015   419,950.00         2       1.75         1350   
9834200885   7/17/2014   360,000.00         4       2.50         2080   
9834200885   4/20/2015   550,000.00         4       2.50         2080   
8062900070    9/9/2014   272,000.00         5       1.50         2550   
8062900070   2/13/2015   369,000.00         5       1.50         2550   
1231000510   9/22/2014   263,000.00         3       1.75         1490   
1231000510    5/4/2015   510,000.00         3       1.75         1490   
6117501820   6/18/2014   250,275.00         2       1.00          790   
6117501820   4/28/2015   435,000.00         2       1.00          790   
2228900270   8/12/2014   215,000.00         2       1.00         1010   
2228900270   2/12/2015   302,000.00         2       1.00         1010   
3630120700   5/13/2014   757,000.00         3       3.25         3190   
3630120700    1/7/2015   765,000.00         3       3.25         3190   
7888000390   6/27/2014   140,000.00         3       1.00         1060   
7888000390    4/1/2015   235,000.00         3       1.00         1060   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                             
6021501535      5000    1.00        0.00  0.00          3      8        1290   
6021501535      5000    1.00        0.00  0.00          3      8        1290   
4139480200     12103    1.00        0.00  3.00          3     11        2690   
4139480200     12103    1.00        0.00  3.00          3     11        2690   
7520000520     12092    1.00         nan  0.00          3      6         960   
7520000520     12092    1.00        0.00  0.00          3      6         960   
3969300030      7134    1.00        0.00  0.00          3      6        1000   
3969300030      7134    1.00        0.00  0.00          3      6        1000   
2231500030     10754    1.00         nan  0.00          5      7        1100   
2231500030     10754    1.00        0.00  0.00          5      7        1100   
8820903380     13579    2.00        0.00  0.00          3      7        2660   
8820903380     13579    2.00        0.00  0.00          3      7        2660   
726049190       7200    1.00        0.00  0.00          4      7        1130   
726049190       7200    1.00         nan  0.00          4      7        1130   
8682262400      4003    1.00        0.00  0.00          3      8        1350   
8682262400      4003    1.00        0.00  0.00          3      8        1350   
9834200885      4080    1.00        0.00  0.00          5      7        1040   
9834200885      4080    1.00        0.00  0.00          5      7        1040   
8062900070      6300    1.00        0.00  0.00          4      7        1560   
8

In most cases, the target value (`'price'`) for the second (last) instance of each group of duplicates seems to be higher than the target value for the first instance(s). However, the values of the features seem to be equal. So we keep those last instances as otherwise the values of the features might be wrong.

In [7]:
houses = houses.loc[~houses.index.duplicated(keep='last')]

## Reordering features

We reorder the features.

In [8]:
houses = houses.reindex(columns=['price', 'date', 'sqft_living', 'sqft_above', 'sqft_basement', 'sqft_lot',
                                 'sqft_living15', 'sqft_lot15', 'bedrooms', 'bathrooms', 'floors', 'yr_built',
                                 'yr_renovated', 'lat', 'long', 'zipcode', 'condition', 'grade', 'view', 
                                 'waterfront'])

## Documentation of features

In [9]:
features = pd.DataFrame([
    {'feature': 'date', 'description': 'date when object was sold', 'type': 'date time'},
    {'feature': 'sqft_living', 'description': 'square footage of interior housing living space',
     'type': 'continuous (numeric)'},
    {'feature': 'sqft_above', 'description': 'square footage of house apart from basement',
     'type': 'continuous (numeric)'},
    {'feature': 'sqft_basement', 'description': 'square footage of basement',
     'type': 'continuous (numeric)'},
    {'feature': 'sqft_lot', 'description': 'square footage of land lot', 'type': 'continuous (numeric)'},
    {'feature': 'sqft_living15',
     'description': 'square footage of interior housing living space for the nearest 15 neighbors',
     'type': 'continuous (numeric)'},
    {'feature': 'sqft_lot15', 'description': 'square footage of land lots of the nearest 15 neighbors',
     'type': 'continuous (numeric)'},
    {'feature': 'bedrooms', 'description': 'number of bedrooms in house', 'type': 'discrete (numeric)'},
    {'feature': 'bathrooms', 'description': 'number of bathrooms in house', 'type': 'discrete (numeric)'},
    {'feature': 'floors', 'description': 'total floors (levels) in house', 'type': 'discrete (numeric)'},
    {'feature': 'yr_built', 'description': 'year when house was built', 'type': 'discrete (numeric)'},
    {'feature': 'yr_renovated', 'description': 'year when house was renovated',
     'type': 'discrete (numeric)'},
    {'feature': 'lat', 'description': 'latitude coordinate of object', 'type': 'continuous (numeric)'},
    {'feature': 'long', 'description': 'longitude coordinate of object', 'type': 'continuous (numeric)'},
    {'feature': 'zipcode', 'description': 'zipcode of area where house is positioned',
     'type': 'nominal (categorical)'},
    {'feature': 'condition', 'description': 'overall condition of object', 'type': 'ordered (categorical)'},
    {'feature': 'grade',
     'description': 'overall grade given to the housing unit, based on King County grading system',
     'type': 'ordered (categorical)'},
    {'feature': 'view', 'description': 'indicator for views of object', 'type': 'ordered (categorical)'},
    {'feature': 'waterfront', 'description': 'indicator whether house lies at a waterfront',
     'type': 'boolean'}
]).astype('string').set_index('feature')

## Summary

In [10]:
houses.sample(5, random_state=0)

price        date  sqft_living  sqft_above  sqft_basement  \
id                                                                          
3226049530 465,000.00   1/22/2015         2010        1290         720.00   
2877102330 772,000.00   5/16/2014         2110        2110           0.00   
8651443420 280,000.00  10/17/2014         1710        1030         680.00   
9521101520 543,000.00  12/12/2014          940         940           0.00   
1026069134 619,000.00   8/25/2014         2560        2560           0.00   

            sqft_lot  sqft_living15  sqft_lot15  bedrooms  bathrooms  floors  \
id                                                                             
3226049530      7264           1510        7326         5       3.00    1.00   
2877102330      3750           1700        5000         4       2.50    2.00   
8651443420      5440           1620        6696         4       2.00    1.00   
9521101520      3864           1440        3956         2       1.00    1.00   
1026069134     43608           3000       54088         3       2.50    2.00   

            yr_built  yr_renovated   lat    long  zipcode  condition  grade  \
id                                                                            
3226049530      1990           nan 47.69 -122.33    98103          3      7   
2877102330      2000          0.00 47.68 -122.36    98117          3      8   
8651443420      1976          0.00 47.37 -122.09    98042          5      8   
9521101520      1918          0.00 47.66 -122.34    98103          4      8   
1026069134      2002          0.00 47.76 -122.03    98077          3      9   

            view  waterfront  
id                            
3226049530  0.00        0.00  
2877102330  0.00        0.00  
8651443420  0.00        0.00  
9521101520  0.00        0.00  
1026069134  0.00        0.00

In [11]:
data_type_info(houses)

dtype  n_unique  p_unique  n_na  p_na
price          float64      3595      0.17     0  0.00
date            object       372      0.02     0  0.00
sqft_living      int64      1034      0.05     0  0.00
sqft_above       int64       942      0.04     0  0.00
sqft_basement  float64       303      0.01   452  0.02
sqft_lot         int64      9776      0.46     0  0.00
sqft_living15    int64       777      0.04     0  0.00
sqft_lot15       int64      8682      0.41     0  0.00
bedrooms         int64        12      0.00     0  0.00
bathrooms      float64        29      0.00     0  0.00
floors         float64         6      0.00     0  0.00
yr_built         int64       116      0.01     0  0.00
yr_renovated   float64        70      0.00  3813  0.18
lat            float64      5033      0.23     0  0.00
long           float64       751      0.04     0  0.00
zipcode          int64        70      0.00     0  0.00
condition        int64         5      0.00     0  0.00
grade            int64        11      0.00     0  0.00
view           float64         5      0.00    63  0.00
waterfront     float64         2      0.00  2353  0.11

In [12]:
features

description  \
feature                                                            
date                                   date when object was sold   
sqft_living      square footage of interior housing living space   
sqft_above           square footage of house apart from basement   
sqft_basement                         square footage of basement   
sqft_lot                              square footage of land lot   
sqft_living15  square footage of interior housing living spac...   
sqft_lot15     square footage of land lots of the nearest 15 ...   
bedrooms                             number of bedrooms in house   
bathrooms                           number of bathrooms in house   
floors                            total floors (levels) in house   
yr_built                               year when house was built   
yr_renovated                       year when house was renovated   
lat                                latitude coordinate of object   
long                              longitude coordinate of object   
zipcode                zipcode of area where house is positioned   
condition                            overall condition of object   
grade          overall grade given to the housing unit, based...   
view                               indicator for views of object   
waterfront          indicator whether house lies at a waterfront   

                                type  
feature                               
date                       date time  
sqft_living     continuous (numeric)  
sqft_above      continuous (numeric)  
sqft_basement   continuous (numeric)  
sqft_lot        continuous (numeric)  
sqft_living15   continuous (numeric)  
sqft_lot15      continuous (numeric)  
bedrooms          discrete (numeric)  
bathrooms         discrete (numeric)  
floors            discrete (numeric)  
yr_built          discrete (numeric)  
yr_renovated      discrete (numeric)  
lat             continuous (numeric)  
long            continuous (numeric)  
zipcode        nominal (categorical)  
condition      ordered (categorical)  
grade          ordered (categorical)  
view           ordered (categorical)  
waterfront                   boolean

## Train-test-split

In [13]:
from sklearn.model_selection import train_test_split

(houses_train, houses_test) = train_test_split(houses, random_state=0)

## Save data set

We save the data set.

In [14]:
houses_train.to_pickle('data/king_county_train_1.pickle')
houses_test.to_pickle('data/king_county_test_1.pickle')
features.to_pickle('data/features_1.pickle')